### PR TITLE
Added `ui:*Template` properties

### LINKED ISSUE #1010: "ui:template" schema option
### Prerequisites

- [x] I have read the [documentation](https://github.com/mozilla-services/react-jsonschema-form/blob/master/README.md#readme);
- [ ] In the case of a bug report, I understand that providing a [SSCCE](http://sscce.org/) example is tremendously useful to the maintainers.

### Description

For more theming control a `"ui:template"` option would be helpful. From what I can tell this would be a fairly simple matter to integrate three places:
 - `SchemaField`
 - `ObjectField`
 - `ArrayField`

This would build on the work in #304.

#### Expected behavior

People could use custom templates to customize layout. This would be the most pertinent for objects and arrays, but can only be useful for leaf nodes too.

```
const uiSchema = {
  "ui:template": GroupedObjectTemplate,
  title: {
    "ui:template": InlineField
  }
  done: {
    "ui:template": ErrorsInTooltip
  },
  comments: {
    "ui:template": InlineArray
  }
};
```

### Alternative 1 - External plugin

Currently as a workaround this behaviour can be (almost) achieved externally. The biggest blocker to making it easier to achieve externally is that the default template objects are not exported.

```
import React from "react";
// TODO: Would be useful to 
import {
  DefaultObjectFieldTemplate,
  DefaultFieldTemplate,
  DefaultArrayItem
} from "./DefaultTemplates";

export function FieldTemplate(props) {
  const Template = props.uiSchema["ui:template"];
  if (
    Template &&
    // TODO: LV: Seems strange that this is required to be plugged in... oh well
    props.schema.type !== "object" &&
    props.schema.type !== "array"
  ) {
    return <Template {...props} />;
  } else {
    return <DefaultFieldTemplate {...props} />;
  }
}

export const ObjectFieldTemplate = defaultOrComponent(
  DefaultObjectFieldTemplate
);
export const ArrayFieldTemplate = defaultOrComponent(DefaultArrayItem);

function defaultOrComponent(DefaultTemplate) {
  return function(props) {
    const Template = props.uiSchema["ui:template"];
    if (Template) {
      console.error("Can't render", Template, props);
      return <Template {...props} />;
    } else {
      return <DefaultTemplate {...props} />;
    }
  };
}
```
And then use that on your `Form`
```
  import * as UiTemplate from "...";
  
  const fm = <Form
    schema={schema}
    uiSchema={uiSchema}
    onChange={log("changed")}
    onSubmit={log("submitted")}
    onError={log("errors")}
    transformErrors={log("transformErrors")}
    liveValidate={true}
    formContext={{
      templates: Templates.GroupTemplates
    }}
    {...UiTemplate}
  >
```

### Alternative 2 - "ui:field"

"ui:field" is very powerful, but it requires too much re-implementation of core behaviour to get right, particularly for things like arrays and objects.

## Recommendation

I would recommend tackling this in phases, getting simpler functionality merged in earlier, and complex functionality merged in later.

*Phase 1* allow templates as objects

 - Before: `const Template = registry.ObjectFieldTemplate || DefaultObjectFieldTemplate;`
 - After: `const Template = uiSchema["ui:template"] || registry.ObjectFieldTemplate || DefaultObjectFieldTemplate;`

*Phase 2* allow templates as strings, add `templates` to registry, allow custom templates map, add default templates to default registry

*Phase 3* themes for other ui libraries

## Resolution

If the maintainers of this repository are on-board, then I can work on a pull request.

### PR DESCRIPTION
**Pull request by loganvolkers**
### Reasons for making this change

Makes customizing templates for a form a more granular experience.

Resolves #1010

### Checklist

* [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://react-jsonschema-form.readthedocs.io/en/latest/#contributing) of the Markdown text I've added
* [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests
  - [ ] I've updated [docs](https://react-jsonschema-form.readthedocs.io/) if needed
  - [x] I've run `npm run cs-format` on my branch to conform my code to [prettier](https://github.com/prettier/prettier) coding style
* [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### CODE REVIEW

**Comment by loganvolkers** [comment]
@epicfaace pull request for adding `ui:template`.

Would be interested in some feedback on the naming convention. Since an array template has both a `SchemaField` template and an `ArrayField` template nested inside, it wasn't possible to have a single `ui:template` property, as it is ambiguous which template needs to be overridden.

Instead I opted to follow the naming convention for overriding fields globally:

- `ui:FieldTemplate` for all fields
- `ui:ArrayFieldTemplate` for the array field subtemplates
- `ui:ObjectFieldTemplate` for the object field subtemplates

I found this the most consistent with the current API. In the future it would be handy if the templating system could be consolidated in a way that a single `ui:template` field would make sense, but that's a substantial non-backwards-compatible undertaking.

This is I think the best compromise between allowing granular overriding of field templates and respecting existing behaviour and API design.

File: package.json [hunk 1]
```diff
@@ -7,8 +7,8 @@
     "build:lib": "rimraf lib && cross-env NODE_ENV=production babel -d lib/ src/",
     "build:dist": "rimraf dist && cross-env NODE_ENV=production webpack --config webpack.config.dist.js",
     "build:playground": "rimraf build && cross-env NODE_ENV=production webpack --config webpack.config.prod.js && cp playground/index.prod.html build/index.html",
-    "cs-check": "prettier -l $npm_package_prettierOptions '{playground,src,test}/**/*.js'",
```
Comment (epicfaace) [inline package.json]: I don't think we need this change?

**Comment by epicfaace** [comment]
> * `ui:FieldTemplate` for all fields
> * `ui:ArrayFieldTemplate` for the array field subtemplates
> * `ui:ObjectFieldTemplate` for the object field subtemplate

@loganvolkers that works. What I'm thinking, though, is what if we have our uiSchema in JSON and want to reference "ui:FieldTemplate", "ui:arrayFieldTemplate", or "ui:ObjectFieldTemplate" through a string, not a React component? The way it's done right now for `ui:widget` and `ui:field` is through the `widgets` and `fields` props on the Form, respectively.

Is the only way to do this to define `fieldTemplates`, `arrayFieldTemplates`, and `objectFieldTemplates` props on the Form as well? That seems like it will clutter the props a lot, especially given that we already have props called `FieldTemplate`, `ArrayFieldTemplate`, and `ObjectFieldTemplate`. Any thoughts?

File: package.json (hunk 1 above)
Comment (loganvolkers) [reply package.json]: These changes to the quotes fix running prettier on Windows.

File: test/FieldTemplate_test.js [hunk 2]
```diff
@@ -19,22 +19,41 @@ describe("FieldTemplate", () => {
...
+        const { node } = createFormComponent({
+          schema: { type: "string" },
+          uiSchema: { "ui:disabled": false },
+          FieldTemplate,
+        });
+        expect(node.querySelectorAll(".disabled")).to.have.length.of(0);
       });
-      expect(node.querySelectorAll(".disabled")).to.have.length.of(1);
     });
+    describe("with template configured in ui:ObjectFieldTemplate", () => {
```
[the remark this answers is not in the document]
Comment (loganvolkers) [reply test/FieldTemplate_test.js]: Feel free to use the "Suggestions" feature of github for these small things :)

```suggestion
    describe("with template configured in ui:FieldTemplate", () => {
```
image

**Comment by loganvolkers** [comment]
>What I'm thinking, though, is what if we have our uiSchema in JSON and want to reference "ui:FieldTemplate", "ui:arrayFieldTemplate", or "ui:ObjectFieldTemplate" through a string 
> then we have a nice prop called templates that we can use for this purpose

I agree with this. `templates` could be added to the `registry` and use the same capabilities that exist for string-referenced widgets and fields.

e.g. for an object field:

```js
<Form
  {...props}
  templates={{"tabs":MyTabTemplate}}
  uiSchema={{"ui:ObjectFieldTemplate":"tabs"}}
/>
```

> and don't let FieldTemplate be customized for array fields

This is a non-starter for me. The goal is to provide UI customization and there is a lot of template-related stuff in the FieldTemplate, especially for Object fields with the new `oneOf` support.

File: package.json (hunk 1 above)
Comment (epicfaace) [reply package.json]: Got it, thanks

File: test/FieldTemplate_test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (epicfaace) [reply test/FieldTemplate_test.js]: Ooh, I always wondered how people did that. Thanks! 😀

**Comment by epicfaace** [comment]
@loganvolkers that sounds good, using `ui:ArrayFieldTemplate`, `ui:FieldTemplate`, and `ui:ObjectFieldTemplate` and then adding support for the `templates` prop in a later PR. For now, can you add the additional tests I mentioned and update the docs? -- and then I think it'll be ready to be merged.

**Comment by epicfaace** [comment]
@loganvolkers will you have time to make these changes? I'd like to merge this PR

**Comment by epicfaace** [comment]
@loganvolkers just bumping this!

**Comment by DavidHenri008** [comment]
@loganvolkers any update on this? This looks really promising.

**Comment by loganvolkers** [comment]
Working on wrapping this up.

I just made some progress on merging the latest to bring this branch up to speed, but there are still some changes requested for tests and docs.

**Comment by loganvolkers** [comment]
@epicfaace can you please review? I believe this should meet your requests.

File: test/ObjectFieldTemplate_test.js [hunk 3]
```diff
@@ -45,33 +45,57 @@ describe("ObjectFieldTemplate", () => {
...
+        "ui:ObjectFieldTemplate": ObjectFieldTemplate,
+      },
+      formData,
+      fields: {
+        TitleField,
+        DescriptionField,
+      },
+    }).node;
+    sharedIts();
   });
```
[the remark this answers is not in the document]
Comment (loganvolkers) [reply test/ObjectFieldTemplate_test.js]: Added

File: test/FieldTemplate_test.js [hunk 4]
```diff
@@ -19,22 +19,41 @@ describe("FieldTemplate", () => {
...
-        FieldTemplate,
+      it("should render with disabled when ui:disabled is falsey", () => {
+        const { node } = createFormComponent({
+          schema: { type: "string" },
+          uiSchema: { "ui:disabled": false, "ui:FieldTemplate": FieldTemplate },
+        });
+        expect(node.querySelectorAll(".disabled")).to.have.length.of(0);
       });
-      expect(node.querySelectorAll(".disabled")).to.have.length.of(0);
     });
```
[the remark this answers is not in the document]
Comment (loganvolkers) [reply test/FieldTemplate_test.js]: Added

File: test/ArrayFieldTemplate_test.js [hunk 5]
```diff
@@ -121,52 +162,72 @@ describe("ArrayFieldTemplate", () => {
...
+        };
+        beforeEach(() => {
+          node = createFormComponent({
+            formData,
+            schema,
+            uiSchema,
+          }).node;
+        });
+        sharedIts();
       });
```
[the remark this answers is not in the document]
Comment (loganvolkers) [reply test/ArrayFieldTemplate_test.js]: Added

File: test/ArrayFieldTemplate_test.js [hunk 6]
```diff
@@ -65,52 +84,74 @@ describe("ArrayFieldTemplate", () => {
...
+
+        beforeEach(() => {
+          node = createFormComponent({
+            formData,
+            schema,
+            uiSchema,
+          }).node;
+        });
+        sharedIts();
       });
```
[the remark this answers is not in the document]
Comment (loganvolkers) [reply test/ArrayFieldTemplate_test.js]: Added

**Comment by epicfaace** [comment]
Thanks @loganvolkers looks good -- can you update the docs?

**Comment by loganvolkers** [comment]
@epicfaace merged latest changes and updated docs. Please take a look.

**Comment by epicfaace** [comment]
Looks good -- thanks!

**Comment by timkindberg** [comment]
> I agree with this. templates could be added to the registry and use the same capabilities that exist for string-referenced widgets and fields.

Was this piece implemented?

**Comment by loganvolkers** [comment]
No, it was not.

**Comment by epicfaace** [comment]
> > I agree with this. `templates` could be added to the registry and use the same capabilities that exist for string-referenced widgets and fields.
> 
> Was this piece implemented?

It's now being tracked in #1387, though.

**Pull request merged**